### PR TITLE
Support simple API names and add `locale` property

### LINKED ISSUE #8: Library Alpha Discussion: Is the API too verbose?
Hey AoG Devs!

Now that the alpha version of the testing library has been out for a few weeks, we'd like to get feedback on the behavior and API design of the library.

The API contains several instances of methods that concatenate more than one word.
`action.setLocale`
`action.startConversation`
`action.startConversationWith`
`action.endConversation`

For most APIs, having simple function names are preferable. In testing libraries, APIs tend to be slightly more verbose. Should it make sense to simplify the set of APIs so that they’d be more like:

`action.locale = ‘en-US’`
`action.start()`
`action.start(‘number genie’)`
`action.cancel()`

### PR DESCRIPTION
**Pull request by yoichiro**
This pull request will fix #8.

# What is this pull request

In this current version, the names of APIs tend to be verbose. Especially, they are the followings:

* `startConversation()`, `startConversationWith()`
* `endConversation()`
* `setLocale()`

In this pull request, these aliases are provided. That is, the following APIs are provided:

* `start()`, `startWith()`
* `cancel()`
* `locale` (used as `action.locale = 'ja-JP'`)

# How to implement them

For adding the `locale` property, I will add a new definition of the `locale` setter method into the `ActionsOnGoogle` class.

For adding the alias methods, I will add the alias method definitions to the `prototype` of the `ActionsOnGoogle` class.

### CODE REVIEW

File: src/actions-on-google.ts [hunk 1]
```diff
@@ -351,4 +355,14 @@ export class ActionsOnGoogle {
             conversation.end()
         })
     }
-}
\ No newline at end of file
+}
+
+// Create some method aliases
+export interface ActionsOnGoogle {
```
Comment (Canain) [inline src/actions-on-google.ts]: what are the pros and cons of aliasing this way versus just writing methods in the classes itself that calls what they alias?

Comment (atulep) [reply src/actions-on-google.ts]: This discussion may help understand reason for aliasing - https://github.com/actions-on-google/actions-on-google-testing-nodejs/issues/8

Comment (Canain) [reply src/actions-on-google.ts]: I mean about comparing aliasing using the prototype rather than explicitly writing methods in the class itself that just call the method they are aliasing.

```javascript
ClassName.prototype.alias = ClassName.prototype.original
```

compared to

```javascript
class ClassName {
  alias(args) {
    this.original(args)
  }
}
```

Comment (yoichiro) [reply src/actions-on-google.ts]: @Canain I think that each approach has pros/cons:

* Add the alias to the prototype of the ActionsOnGoogle
  * Pros: The count of each alias method call doesn't increase.
  * Cons: Each definition of alias methods isn't placed in the ActionsOnGoogle class. Therefore, if using typedoc to generate API reference docs, the documents of the alias methods may not be generated.
* Define as an instance method of the ActionsOnGoogle
  * Pros: Each definition of alias methods is placed in the ActionsOnGoogle class. Therefore, if using typedoc to generate API reference docs, the documents of the alias methods can be generated as same as the original method.
  * Cons: The count of each alias method call increases one time ( `start()` -> `startConversation()` ).

If we should think about performance as high priority, we should select the approach of using the prototype. However, I guess that the performance is not high priority in this case. I wrote the code using the prototype, because Nick mentioned about the approach which defines alias methods without adding them as the method of the ActionsOnGoogle class at [#8](https://github.com/actions-on-google/actions-on-google-testing-nodejs/issues/8#issuecomment-403571926). But, maybe, it would be better that we select a simple approach rather than getting a little bit performance.

As the result, I intend to rewrite the code as a new method adding to the ActionsOnGoogle class. How about you?

Comment (Canain) [reply src/actions-on-google.ts]: I think if we are going to support aliases, then explicitly declaring them as methods in the class is probably the better approach like you mentioned.

While it's more verbose, the declaration is more explicit and using standard OOP syntax meaning that it's more easily understood and parsable by tooling. The performance decrease shouldn't be important for a library like this.

Ideally, it would be nice to only support one name for when the library is stable, but it may not be feasible to solve for everyone.

**Comment by yoichiro** [comment]
@Canain I have changed the alias approach to adding methods simply. I would like you to check the following:

* `start()` alias method: https://github.com/actions-on-google/actions-on-google-testing-nodejs/pull/11/files#diff-59d7d620304bea33809f81e10d79406eR287
* `startWith()` alias method: https://github.com/actions-on-google/actions-on-google-testing-nodejs/pull/11/files#diff-59d7d620304bea33809f81e10d79406eR298
* `cancel()` alias method: https://github.com/actions-on-google/actions-on-google-testing-nodejs/pull/11/files#diff-59d7d620304bea33809f81e10d79406eR306

**Comment by Canain** [comment]
@Fleker can you take a look and see if it works for you?

**Comment by Fleker** [comment]
I do like the more concise methods, although I am hesitant to start creating a bunch of aliases for methods before we even get to a v1 as it'll end up cluttering the API.

Maybe we should use these new methods as the default way and deprecate the current versions. We can accept the aliasing for now, with a plan to remove those verbose method names in a v1 release.

**Comment by yoichiro** [comment]
@Fleker @Canain  Thank you for the confirmation. Ok, I intend to change my code to the following:

* Move the implementation code to each new method. And, change the code of each original method to call the corresponding new method.
* Add comments with typedoc for classes and their properties and methods. And, put the `@deprecated` annotation on the original methods to represent that the original methods will be removed.

I would like you to review again after I change and commit them.

**Comment by yoichiro** [comment]
@Fleker @Canain Could you review the following commits?

* dbe1a9d - Move the implementation code to each new method. And, change the code of each original method to call the corresponding new method.
* 946b4bd - Introduce a typedoc. And, put the @deprecated annotation on the original methods to represent that the original methods will be removed.

File: src/actions-on-google.ts [hunk 2]
```diff
@@ -260,7 +260,14 @@ export class ActionsOnGoogle {
         return client
     }
 
+    /**
+     * @deprecated This will be removed at releasing a version 1.
```
Comment (Fleker) [inline src/actions-on-google.ts]: I would rephrase it as
```
/**
 * @deprecated This will be removed in future releases. Use `action.locale` instead
 */
```

File: package.json [hunk 3]
```diff
@@ -13,7 +13,9 @@
         "clean": "rm -rf dist",
```
Comment (Fleker) [inline package.json]: Should `yarn clean` also run `yarn docs:clean`?

File: README.md [hunk 4]
```diff
@@ -70,6 +70,16 @@ action.startTest('Facts about Google - direct cat path', action => {
...
+## Supported features
+
+This library provides the following features to control your conversation:
+
+* `action.startConversation()`, `action.start()` - Start your conversation with your action using "my test app".
+* `action.startConversationWith()`, `action.startWith()` - Start your conversation with your action using the specified action name.
+* `action.send()` - Send some phrase to your action.
+* `action.endConversation()`, `action.cancel()` - End your conversation. This library says "cancel".
+* `action.setLocale()`, `action.locale` - Set a locale for your conversation.
+
```
Comment (Fleker) [inline README.md]: We should probably move `action.location` up to this section of the README, then remove the **Additional features** section at the bottom.

**Comment by yoichiro** [comment]
@Fleker I have just changed codes you mentioned. Could you review diffs 3086897 5d3dbd4 1336a5e again?

**Comment by yoichiro** [comment]
@atulep Could you confirm this?

**Comment by atulep** [comment]
First of all, thank you so much for making the changes.

Couple of things:
1. Can you also change examples/facts-about-google-ts to use simplified interface, since you modified both of the other examples?
2. Can you change the src/test/ to use updated API?

**Comment by yoichiro** [comment]
@atulep Thank you for the reviewing. Both answers of the questions are yes. I do them.

**Comment by yoichiro** [comment]
@atulep Could you review their changes below?

* Can you also change examples/facts-about-google-ts to use simplified interface, since you modified both of the other examples? -> b12972c
* Can you change the src/test/ to use updated API? -> e2d617d

**Comment by yoichiro** [comment]
@atulep Thank you for the reviewing!

**Comment by yoichiro** [comment]
@Fleker @Canain Could you do this final judgement?

**Comment by Fleker** [comment]
Yeah this looks good to me.

**Comment by Canain** [comment]
Looks good to me.

**Comment by yoichiro** [comment]
@Canain @Fleker @atulep Thanks, guys. Can somebody merge this to the master branch? :)

**Comment by Fleker** [comment]
I ran the tests and everything is good. We'll do the merge and release on Monday, so as not to break things over the weekend. :grin:

**Comment by yoichiro** [comment]
@Fleker Ok, have an enjoy your weekend!
